File: android/app/src/main/java/com/izzatismail/midtrim/domain/usecase/ValidateTrimDurationUseCase.kt
package com.izzatismail.midtrim.domain.usecase

import com.izzatismail.midtrim.domain.error.ValidateTrimDurationError

class ValidateTrimDurationUseCase {
    private val freeTierAllowed = setOf(1.0, 2.0, 3.0)
    private val paidTierRange = 1.0..5.0

    fun execute(trimDuration: Double, isPaidUser: Boolean) {
        if (trimDuration !in paidTierRange) {
            if (trimDuration < 1.0) {
                throw ValidateTrimDurationError.BelowMinimumDuration(trimDuration)
            }
            throw ValidateTrimDurationError.ExceedsMaxDuration(trimDuration)
        }
        if (!isPaidUser && trimDuration !in freeTierAllowed) {
            throw ValidateTrimDurationError.FreeTierRejectsCustomDuration(trimDuration)
        }
    }

    fun isAllowed(trimDuration: Double, isPaidUser: Boolean): Boolean {
        return try {
            execute(trimDuration, isPaidUser)
            true
        } catch (_: ValidateTrimDurationError) {
            false
        }
    }
}

File: ios/MidTrim/MidTrim/Domain/UseCases/ValidateTrimDurationUseCase.swift
import Foundation

struct ValidateTrimDurationUseCase {
    private let freeTierAllowed: Set<Double> = [1, 2, 3]
    private let paidTierRange: ClosedRange<Double> = 1.0...5.0

    func execute(trimDuration: Double, isPaidUser: Bool) throws {
        guard paidTierRange.contains(trimDuration) else {
            if trimDuration < 1.0 {
                throw ValidateTrimDurationError.belowMinimumDuration(trimDuration)
            }
            throw ValidateTrimDurationError.exceedsMaxDuration(trimDuration)
        }

        if !isPaidUser {
            guard freeTierAllowed.contains(trimDuration) else {
                throw ValidateTrimDurationError.freeTierRejectsCustomDuration(trimDuration)
            }
        }
    }

    func isAllowed(trimDuration: Double, isPaidUser: Bool) -> Bool {
        (try? execute(trimDuration: trimDuration, isPaidUser: isPaidUser)) != nil
    }
}
